execution/tvf/watermark_generator.go

```diff
@@ -0,0 +1,107 @@
+package tvf
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/cube2222/octosql"
+	"github.com/cube2222/octosql/docs"
+	"github.com/cube2222/octosql/execution"
+	"github.com/cube2222/octosql/streaming/storage"
+)
+
+type WatermarkGenerator struct {
+	source    execution.Node
+	timeField octosql.VariableName
+}
+
+func NewWatermarkGenerator(source execution.Node, timeField octosql.VariableName) *WatermarkGenerator {
+	return &WatermarkGenerator{
+		source:    source,
+		timeField: timeField,
+	}
+}
+
+func (w *WatermarkGenerator) Document() docs.Documentation {
+	panic("implement me")
+}
+
+func (w *WatermarkGenerator) Get(ctx context.Context, variables octosql.Variables, streamID *execution.StreamID) (execution.RecordStream, *execution.ExecOutput, error) {
+	tx := storage.GetStateTransactionFromContext(ctx)
+	sourceStreamID, err := execution.GetSourceStreamID(tx.WithPrefix(streamID.AsPrefix()), octosql.MakePhantom())
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get source stream ID")
+	}
+
+	source, _, err := w.source.Get(ctx, variables, sourceStreamID) // we don't need execOutput here since we become new watermark source
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get source")
+	}
+
+	ws := &WatermarkGeneratorStream{
+		source:    source,
+		timeField: w.timeField,
+	}
+
+	return ws, execution.NewExecOutput(ws), nil // watermark generator stream now indicates new watermark source
+}
+
+type WatermarkGeneratorStream struct {
```

Comment:
This watermark generator currently uses the maximum record seen as a watermark. This is a special case of the watermark generator we wanted to support which uses the maximum - some interval. For example, the watermark is the max record seen minus 10 seconds (so we expect lack of order no bigger than 10 seconds in our input stream)

Reply:
So please just modify this one to accept such a parameter.